### PR TITLE
Add functions to functions.yaml

### LINKED ISSUE #1828: functions.yaml is incomplete
Not all public functions are listed in `functions.yaml`. This issue is to:

 - Identify what is missing.
 - Add all missing functions to `functions.yaml`

### PR DESCRIPTION
**Pull request by GroteGnoom**
To fix #1828

### CODE REVIEW

File: interfaces/functions.yaml [hunk 1]
```diff
@@ -171,6 +174,13 @@ igraph_realize_degree_sequence:
         OUT GRAPH graph, VECTOR_INT out_deg, OPTIONAL VECTOR_INT in_deg=NULL,
         EDGE_TYPE_SW allowed_edge_types=SIMPLE, REALIZE_DEGSEQ_METHOD method=SMALLEST
 
+igraph_circulant:
+    PARAMS: OUT GRAPH graph, INTEGER n, VECTOR_INT shifts, BOOLEAN directed
+
+igraph_weighted_sparsemat:
+    PARAMS:
+        OUT GRAPH graph, SPARSEMAT A, BOOLEAN directed, CSTRING attr, BOOLEAN loops=False
```
Comment (ntamas) [inline interfaces/functions.yaml:182]: Note to ourselves: this reflects the current state correctly, but we will probably have to update `igraph_weighted_sparsemat` soon to be consistent with `igraph_weighted_adjacency` in the sense that it does not take an attribute _name_ but a vector (in which the weights will be stored). See `CHANGELOG.md` for the reasons.

File: interfaces/functions.yaml [hunk 2]
```diff
@@ -572,6 +608,19 @@ igraph_personalized_pagerank:
...
         options ON algo
 
+igraph_personalized_pagerank_vs:
+    PARAMS: |-
+        GRAPH graph, PAGERANKALGO algo=PRPACK,
+        OUT VERTEX_QTY vector, OUT REAL value,
+        VERTEXSET vids=ALL, BOOLEAN directed=True,
+        REAL damping=0.85,
+        VERTEXSET reset_vids,
+        OPTIONAL EDGEWEIGHTS weights,
```
Comment (ntamas) [inline interfaces/functions.yaml:618]: Default value of `weights` should be `NULL`

File: interfaces/functions.yaml [hunk 3]
```diff
@@ -917,6 +966,19 @@ igraph_average_local_efficiency:
         BOOLEAN directed=True, NEIMODE mode=ALL
     DEPS: weights ON graph
 
+igraph_sort_vertex_ids_by_degree:
+    PARAMS: |-
+        GRAPH graph, OUT VERTEX_INDICES outvids, VERTEXSET vids,
+        NEIMODE mode=OUT, BOOLEAN loops=False, ORDER order, BOOLEAN only_indices=False
```
Comment (ntamas) [inline interfaces/functions.yaml:972]: I think that we can reasonably assume ascending order as the default value for `order`

File: interfaces/functions.yaml [hunk 4]
```diff
@@ -1011,6 +1073,13 @@ igraph_bipartite_game_gnm:
         INTEGER n1, INTEGER n2, INTEGER m, BOOLEAN directed,
         NEIMODE mode
 
+igraph_bipartite_game:
+    PARAMS: |-
+        OUT GRAPH graph, OPTIONAL OUT VECTOR_BOOL types,
+        ERDOS_RENYI_TYPE type, INTEGER n1, INTEGER n2, REAL p,
```
Comment (ntamas) [inline interfaces/functions.yaml:1079]: Let's add defaults for `p` and `m` here; `p` is ignored for G(n,m) graphs and `m` is ignored for G(n,p) graphs so it does not make sense to ask the user to specify both.

Similarly, we should provde defaults for `directed` and `mode`

File: interfaces/functions.yaml [hunk 5]
```diff
@@ -1090,6 +1159,12 @@ igraph_maximal_cliques:
     PARAMS: GRAPH graph, OUT VERTEXSET_LIST res, INTEGER min_size=0, INTEGER max_size=0
     DEPS: res ON graph
 
+igraph_maximal_cliques_subset:
+    PARAMS: |-
+        GRAPH graph, VECTOR_INT subset, OUT VECTOR_INT_LIST res,
```
Comment (ntamas) [inline interfaces/functions.yaml:1164]: `subset` should be of type `VERTEX_INDICES` to make sure that R translates between 0-based and 1-based indices when generating code.

Similarly, `res` should be `VERTEXSET_LIST`

File: interfaces/functions.yaml [hunk 6]
```diff
@@ -1090,6 +1159,12 @@ igraph_maximal_cliques:
...
+        OUT INTEGER no, OUTFILE outfile, INTEGER min_size, INTEGER max_size
+    DEPS: res ON graph
```
Comment (ntamas) [inline interfaces/functions.yaml:1166]: `subset` needs to depend on `graph` if we change its type to `VERTEX_INDICES`

File: interfaces/functions.yaml [hunk 7]
```diff
@@ -1742,10 +1842,18 @@ igraph_compose:
         OUT INDEX_VECTOR edge_map1, OUT INDEX_VECTOR edge_map2
     DEPS: edge_map1 ON g1, edge_map2 ON g2
 
+igraph_induced_subgraph_map:
+    PARAMS: |-
+        GRAPH graph, OUT GRAPH res, VERTEXSET vids, SUBGRAPH_IMPL impl, OPTIONAL OUT VECTOR_INT map,
+        OPTIONAL OUT VECTOR_INT invmap
```
Comment (ntamas) [inline interfaces/functions.yaml:1848]: `vids` need to depend on `res` here

File: interfaces/functions.yaml [hunk 8]
```diff
@@ -1766,6 +1874,14 @@ igraph_maxflow_value:
 igraph_mincut_value:
     PARAMS: GRAPH graph, OUT REAL res, EDGE_CAPACITY capacity
 
+igraph_st_mincut:
+    PARAMS: |-
+        GRAPH graph, OUT REAL value, OUT EDGESET cut, OUT VERTEXSET partition1, OUT VERTEXSET partion2,
```
Comment (ntamas) [inline interfaces/functions.yaml:1879]: Typo here: `partion2` should be `partition2`

File: interfaces/functions.yaml [hunk 9]
```diff
@@ -1766,6 +1874,14 @@ igraph_maxflow_value:
...
+        VERTEX source, VERTEX target,
+        EDGE_CAPACITY capacity
+    DEPS:
+        source ON graph, target ON graph
```
Comment (ntamas) [inline interfaces/functions.yaml:1883]: `capacity` needs to depend on `graph`

File: interfaces/types.yaml [hunk 10]
```diff
@@ -46,6 +46,10 @@ CSTRING:
     # A null-terminated immutable C string
     CTYPE: const char*
 
+CSTRING_P:
```
Comment (ntamas) [inline interfaces/types.yaml:49]: This is not needed; declaring a parameter as `OUT CSTRING` will add an extra level of pointer indirection automatically.

File: interfaces/functions.yaml [hunk 11]
```diff
@@ -2174,9 +2343,76 @@ igraph_vertex_coloring_greedy:
...
+
+igraph_expand_path_to_pairs:
+    PARAMS: INOUT VERTEX_INDICES path
+
+#######################################
+# Meta info
+#######################################
+
+igraph_version:
+    PARAMS: CSTRING_P version_string, OPTIONAL OUT INT major, OPTIONAL OUT INT minor, OPTIONAL OUT INT subminor
```
Comment (ntamas) [inline interfaces/functions.yaml:2417]: `version_string` needs to be an `OPTIONAL OUT CSTRING`

File: interfaces/functions.yaml (hunk 7 above)
Comment (ntamas) [reply interfaces/functions.yaml:1848]: Nope, actually, it needs to depend on `graph`

**Comment by szhorvat** [comment]
There were some here that should probably never be exposed to a high-level language such as `igraph_has_attribute_table`, but I guess they don't hurt either.

**Pull request merged**